### PR TITLE
Make cuDNN a runtime dependency rather than a link time dependency

### PR DESCRIPTION
**Pull request by umar456**
* Load cuDNN at runtime
* Take into account different versions of cuDNN
* Add support for older cuDNN versions (Tested with 4, 6 and 7)

### CODE REVIEW

File: src/backend/common/DependencyModule.hpp [hunk 1]
```diff
@@ -42,7 +42,7 @@ class DependencyModule {
     /// Returns true if the module was successfully loaded
     bool isLoaded();
 
-    /// Returns true if the module was successfully loaded
+    /// Returns true if all of the symbols for the module were loaded
```
Comment (9prady9) [inline src/backend/common/DependencyModule.hpp]: :+1:

File: src/backend/common/util.hpp [hunk 2]
```diff
@@ -17,5 +17,13 @@
 std::string getEnvVar(const std::string &key);
 
 // Dump the kernel sources only if the environment variable is defined
-static const char* saveJitKernelsEnvVarName = "AF_JIT_KERNEL_TRACE";
 void saveKernel(const std::string& funcName, const std::string& jit_ker, const std::string& ext);
+namespace {
+const char* saveJitKernelsEnvVarName = "AF_JIT_KERNEL_TRACE";
+
+std::string int_version_to_string(int version) {
```
Comment (9prady9) [inline src/backend/common/util.hpp]: A comment on intended usage is useful. Are the values in the formulation (1000 , 10) always fixed ?

File: src/backend/cuda/CMakeLists.txt [hunk 3]
```diff
@@ -525,6 +524,7 @@ target_include_directories (afcuda
     ${CMAKE_CURRENT_SOURCE_DIR}/kernel
     ${CMAKE_CURRENT_SOURCE_DIR}/jit
     ${CMAKE_CURRENT_BINARY_DIR}
+    #${CUDNN_INCLUDE_DIR}
```
Comment (9prady9) [inline src/backend/cuda/CMakeLists.txt]: Interesting, why is this include commented instead of removal ?

File: src/backend/cuda/convolve.cpp [hunk 4]
```diff
@@ -39,14 +39,20 @@ template<>
...
     return CUDNN_DATA_DOUBLE;
 }
+
+#if CUDNN_MAJOR >= 6
 template<>
 cudnnDataType_t getCudnnDataType<int>() {
     return CUDNN_DATA_INT32;
 }
+
+#elif CUDNN_MAJOR >= 6
```
Comment (9prady9) [inline src/backend/cuda/convolve.cpp]: Why are the else and if conditionals same ?

File: src/backend/cuda/cudnn.hpp [hunk 5]
```diff
@@ -10,22 +10,136 @@
...
+
+
+// cuDNN Wrappers
+//
+// cuDNN deprecates and releases function names often between releases. in order
+// to prevent locking arrayfire versions to specific cuDNN versions, we wrap all
+// cuDNN calls so that the main codebase is not full of ifdefs. The Following
+// functions are wrappers around cuDNN functions that abstract out the version
+// differences between older versions of cuDNN.
+//
```
Comment (9prady9) [inline src/backend/cuda/cudnn.hpp]: How do we handle this inside the wrapper call or the cuDNN module loader ? There doesn't seem to be any guards inside cuDNN module loader based on the cuDNN versions.

**Comment by umar456** [comment]
Addressed all feedback

File: src/backend/cuda/convolve.cpp [hunk 6]
```diff
@@ -39,14 +39,20 @@ template<>
...
     return CUDNN_DATA_DOUBLE;
 }
+
+#if CUDNN_MAJOR >= 6
 template<>
 cudnnDataType_t getCudnnDataType<int>() {
     return CUDNN_DATA_INT32;
 }
+
+#elif CUDNN_MAJOR > 7 || (CUDNN_MAJOR >= 7 && CUDNN_MINOR >= 1)
```
Comment (9prady9) [inline src/backend/cuda/convolve.cpp]: We will never reach the code-path under this condition. I think you were aiming for a separate `#if` here.

File: src/backend/cuda/cudnn.cpp [hunk 7]
```diff
@@ -27,14 +27,169 @@ const char *errorString(cudnnStatus_t err) {
...
+            convDesc, pad_h, pad_w, u, v, upscalex, upscaley, mode);
+#endif
 }
+
+cudnnStatus_t cudnnSetFilter4dDescriptor(cudnnFilterDescriptor_t filterDesc,
+                                         cudnnDataType_t dataType,
+                                         cudnnTensorFormat_t format, int k,
+                                         int c, int h, int w) {
+    int major;
+    tie(major, std::ignore, std::ignore) = getCudnnPlugin().getVersion();
```
Comment (9prady9) [inline src/backend/cuda/cudnn.cpp]: these two lines can be moved into the `#if CUDNN_MAJOR >= 6` arm.

File: src/backend/cuda/cudnn.cpp [hunk 8]
```diff
@@ -27,14 +27,172 @@ const char *errorString(cudnnStatus_t err) {
...
+#if CUDNN_VERSION >= 6000
+    int version = getCudnnPlugin().cudnnGetVersion();
+    if (version >= 6000) {
+        return getCudnnPlugin().cudnnSetFilter4dDescriptor(filterDesc, dataType,
+                                                           format, k, c, h, w);
+    } else if (version == 4000) {
+        return getCudnnPlugin().cudnnSetFilter4dDescriptor_v4(
+            filterDesc, dataType, format, k, c, h, w);
+    }
+    CUDA_NOT_SUPPORTED(
```
Comment (9prady9) [inline src/backend/cuda/cudnn.cpp]: :+1:

File: src/backend/cuda/convolve.cpp [hunk 9]
```diff
@@ -39,14 +39,21 @@ template<>
...
     return CUDNN_DATA_DOUBLE;
 }
+
+#if CUDNN_VERSION >= 6000
 template<>
 cudnnDataType_t getCudnnDataType<int>() {
     return CUDNN_DATA_INT32;
 }
+
+#if CUDNN_VERSION >= 7100
```
Comment (9prady9) [inline src/backend/cuda/convolve.cpp]: :+1:

**Comment by 9prady9** [comment]
Strided convolution tests are failing on k5000.

File: src/backend/common/DependencyModule.cpp [hunk 10]
```diff
@@ -45,14 +47,42 @@ namespace common {
...
+            AF_TRACE("Unable to open {}", plugin_file_name);
+        }
     }
 }
 
+DependencyModule::DependencyModule(const vector<string> plugin_base_file_name,
+                                   const vector<string> suffixes,
+                                   const vector<string> paths)
+    : handle(nullptr), logger(common::loggerFactory("platform")) {
+    UNUSED(paths);
```
Comment (9prady9) [inline src/backend/common/DependencyModule.cpp]: Must be a left over from copy-paste.

**Comment by umar456** [comment]
The ASAN failure is because we are now dynamically loading cuDNN and we close it before we exit the executable. The close operation unloads all of the symbols and the leak sanitizer cannot find the symbols. You can get around this issue by LD_PRELOADing the library and address sanitizer and the leaks go away.

**Pull request merged**